k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -52,60 +57,75 @@ class EndpointsNamer(idPrefix: Path, mkApi: String => NsApi) extends Namer {
 
   private[this] object Ns {
     private[this] var caches: Map[String, NsCache] = Map.empty
+    // XXX once a namespace is watched, it is watched forever.
+    private[this] var _watches = Map.empty[String, Activity[Closable]]
+    def watches = _watches
+
+    /**
+      * Returns an Activity backed by a Future.  The resultant Activity is pending until the
+      * original future is satisfied.  When the Future is successful, the Activity becomes
+      * an Activity.Ok with a fixed value from the Future.  If the Future fails, the Activity
+      * becomes an Activity.Failed and the Future is retried with the given backoff schedule.
+      * Therefore, the legal state transitions are:
+      *
+      * Pending -> Ok
+      * Pending -> Failed
+      * Failed -> Failed
+      * Failed -> Ok
+      */
+    private[this] def retryToActivity[T](go: => Future[T]): Activity[T] = {
+      val state = Var(Activity.Pending: Activity.State[T])
+      _retryToActivity(backoff, state)(go)
+      Activity(state)
+    }
+
+    private[this] def _retryToActivity[T](
+      remainingBackoff: Stream[Duration],
+      state: Var[Activity.State[T]] with Updatable[Activity.State[T]] = Var[Activity.State[T]](Activity.Pending)
+    )(go: => Future[T]): Unit = {
+      val _ = go.respond {
+        case Return(t) =>
+          state() = Activity.Ok(t)
+        case Throw(e) =>
+          state() = Activity.Failed(e)
+          remainingBackoff match {
+            case delay #:: rest =>
+              Future.sleep(delay).onSuccess { _ => _retryToActivity(rest, state)(go) }
+            case Stream.Empty =>
+          }
+      }
+    }
 
     def get(name: String): NsCache = synchronized {
       caches.get(name) match {
         case Some(ns) => ns
         case None =>
-          val ns = mkNs(name)
+          val ns = new NsCache(name)
+          val closable = retryToActivity(watch(name, ns))
```

Comment:
might be slightly clearer as `retryToActivity { watch(name, ns) }` to make it obvious that this is a closure